### PR TITLE
perf(amd): add skip-softmax attention with deferred V load and dynamic scheduling to Gluon MHA prefill (gfx950)

### PR DESCRIPTION
**Pull request by hiroaki-amd**
## Summary

- Add skip-softmax attention (BLASST, arXiv:2512.12087) to the Gluon MHA prefill kernel on gfx950, controlled by `skip_softmax_threshold`. A K/V block is skipped only when every row in the query tile has `exp(block_max_score - running_max) < skip_softmax_threshold`, where `running_max` covers the preceding blocks and excludes the block being tested. If even one row misses, the block runs normally and every row gets the ordinary update, bit-for-bit as if skipping were off. Skipping a block skips the online-softmax update, the `P@V` matmul, and V's load.
- `skip_softmax_threshold=0.0` is the default and is exact dense attention, bit-identical to the existing kernel. A nonzero threshold has to be calibrated per model and per sequence length.
- Add `defer_v_load`: V's HBM->LDS load is issued only for blocks that are not skipped. Causal main loop only; the launcher forces it off for the sliding-window kernel. It changes when V is loaded, never which blocks are skipped, so output is bit-identical to `defer_v_load=False`.
- Replace the static workgroup-to-head assignment with a global atomic work counter when skipping is on, for the causal kernel; the sliding-window kernel keeps the static schedule. Block sparsity varies sharply between heads of the same layer, and with one head pinned per workgroup the layer runs at close to worst-head speed. The counter changes no results, only which workgroup runs which tile; output is bit-identical with it disabled.
- Dispatch is gated by a `support_skip_softmax` kernel trait; a nonzero threshold on an unsupported kernel raises `NoKernelFoundError` rather than silently running dense. `defer_v_load` is not exposed on `mha_prefill()` and is reachable only through `gluon_mha_prefill_gfx950()`.

## Performance

Qwen3-8B, RULER, all 13 tasks, ctx=32768, `skip_softmax_threshold=0.03`, per-layer attention timed over 36 layers against dense measured in the same run on MI350X:

```
  skipping                  1.168x
  + defer_v_load            1.303x
```

Mean block sparsity is 55.2%. Both figures are the mean of the per-layer speedups. With skipping off the counter costs 2.0%, so it is enabled only when `skip_softmax_threshold > 0`.

## Accuracy

Qwen3-8B, RULER, 13 tasks x 50 samples (650 prompts), ctx=32768, `skip_softmax_threshold=0.03`, over a full forward pass with each layer feeding its approximated output to the next:

```
  13-task aggregate    91.58%  ->  91.72%    (+0.15 pts)
```

11 of 13 tasks are unchanged, two lose (`fwe` -2.0, `cwe` -0.6) and three gain (`niah_multivalue` +0.5, `qa_1` +2.0, `qa_2` +2.0). At 50 samples per task a 2-percentage-point move is one prompt, so the aggregate is within the resolution of the measurement. We do not claim an improvement, only that no degradation is measurable at this threshold.

## Test plan

- [x] `pre-commit run --all-files` clean
- [x] `test/ops/test_mha_prefill_skip_softmax.py` (30 tests) on MI350X: threshold range, sliding window, attention sinks, returned LSE
- [x] `test/ops/test_mha_prefill_defer_v_load.py` (34 tests), of which 21 assert `defer_v_load=True` is bit-identical to `False`
- [x] `test/ops/test_mha_prefill_skip_softmax_update.py` (36 tests): the block-level skip rule (a block is skipped only when every row qualifies) and that skipping does change the output once a block qualifies
- [x] `test/ops/test_mha_prefill_dynamic_sched.py` (27 tests): coverage against dense at a threshold too small to skip anything, determinism over repeated launches despite varying atomic arrival order, GQA ratios 1/2/4/8, a ragged batch, and the sliding-window kernel being unaffected by threshold beyond ordinary reduction-order noise (it keeps the static schedule regardless of threshold)
- [x] `test/ops/test_attention_mha_prefill_skip_softmax.py` (5 registry tests): dispatch and the `NoKernelFoundError` path
- [x] Existing `test/ops/test_attention.py` and `test_attention_rel_mha.py` still pass

### CODE REVIEW

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/attention/mha/prefill.py [hunk 1]
```diff
@@ -1066,7 +1238,32 @@ def gluon_mha_prefill_gfx950(
     sinks: torch.Tensor | None = None,
     return_lse: bool = False,
     softmax_scale: float | None = None,
+    skip_softmax_threshold: float = 0.0,
+    defer_v_load: bool = False,
```
Comment (raikonenfnu) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/attention/mha/prefill.py]: Do we ever want to **not** defer_v_load if we are skipping softmax?

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py [hunk 2]
```diff
@@ -170,6 +170,7 @@ def mha_prefill(
     sinks: torch.Tensor | None = None,
     return_lse: bool = False,
     softmax_scale: float | None = None,
+    skip_softmax_threshold: float = 0.0,
```
Comment (raikonenfnu) [inline tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py]: Similar to how it is done in SG, I think we hsould also add a skip softmax option in server args https://github.com/lightseekorg/tokenspeed/blob/4d601173de1dd1bceed6c5cfe7796930508231c3/python/tokenspeed/runtime/utils/server_args.py#L98.

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/attention/mha/prefill.py (hunk 1 above)
Comment (hiroaki-amd) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/attention/mha/prefill.py]: Good question, and the answer is yes. Fixed in `594dbb71`: `defer_v_load` is no longer a caller-facing switch, it's now derived internally as `enable_skip_softmax and not is_sliding`.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py (hunk 2 above)
Comment (hiroaki-amd) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py]: Added in `a856f589`: `--skip-softmax-threshold` is now a `ServerArgs` field, wired through `MHAConfig` into `MHAAttnBackend`, following the same pattern as `--attention-backend`.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py [hunk 3]
```diff
@@ -207,6 +221,8 @@ def mha_prefill(
         "support_sinks": sinks is not None,
         "return_lse": return_lse,
     }
+    if skip_softmax_threshold > 0.0:
+        traits["support_skip_softmax"] = True
```
[the remark this answers is not in the document]
Comment (hiroaki-amd) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py]: Good catch, fixed in `2e23c463`: added the missing `support_skip_softmax` trait declaration to the Ascend entry point.

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/attention/mha/prefill.py [hunk 4]
```diff
@@ -367,24 +381,48 @@ def softmax(self, qk, m_i, l_i, acc):
...
+        # Skip softmax: a row can be skipped when its max score falls more
+        # than log2_threshold below the running max so far (m_i). We compare
+        # against m_i rather than m_new (which already includes this block)
+        # because using m_new would make the difference 0, and every row
+        # would look skippable, on any row where this block sets the new
+        # max. We only skip the whole block when every row in it can be
+        # skipped.
+        if cfg.ENABLE_SKIP_SOFTMAX:
+            skip = (row_max - m_i) * cfg.SM_SCALE < self.log2_threshold
+            all_skip = gl.sum(skip.to(gl.int32), axis=0) == cfg.BLOCK_M
```
[the remark this answers is not in the document]
Comment (hiroaki-amd) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/attention/mha/prefill.py:399]: Fixed in `d09ad975`: padded query rows no longer participate in the per-row skip vote.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hiroaki-amd) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py]: Kept it as a defaulted optional kwarg, matching `logit_cap`, `sinks`, and `window_left` on the same functions. Making it required would touch call sites well beyond this PR's scope (e.g. every `ServerArgs`/`MHAConfig` test fixture across the runtime test suite, most of them unrelated to skip-softmax), for a feature that's meant to be opt-in and leave dense attention untouched by default.

File: python/tokenspeed/runtime/layers/attention/backends/paged/mha.py [hunk 5]
```diff
@@ -393,6 +395,7 @@ def _forward_prefill(
             window_left=layer.sliding_window_size,
             logit_cap=layer.logit_cap,
             sinks=sinks,
+            skip_softmax_threshold=self.skip_softmax_threshold,
```
[the remark this answers is not in the document]
Comment (hiroaki-amd) [reply python/tokenspeed/runtime/layers/attention/backends/paged/mha.py:404]: Good catch. Fixed in `1e020a7e`: documented in the help text that the threshold only reaches the gfx950 prefill kernel on requests with no cached prefix, so FP8/MXFP8 (which always uses the extend path) silently ignores it. Adding skip-softmax support to the extend path itself is out of scope for this PR.

File: python/tokenspeed/runtime/utils/server_args.py [hunk 6]
```diff
@@ -1630,6 +1633,19 @@ def add_cli_args(parser: argparse.ArgumentParser):
...
+            type=float,
+            default=ServerArgs.skip_softmax_threshold,
+            help="BLASST skip-softmax sparsity threshold for the gluon MHA "
+            "prefill kernel (gfx950 only). A K/V block is skipped only when "
+            "every row in the query tile has exp(block_max_score - "
+            "running_max) below this threshold. 0.0 (default) is exact "
+            "dense attention; the skip rate for a given threshold must be "
+            "calibrated per model and sequence length. Ignored by "
+            "attention backends that do not declare gluon skip-softmax "
+            "support.",
```
[the remark this answers is not in the document]
Comment (hiroaki-amd) [reply python/tokenspeed/runtime/utils/server_args.py]: Good catch. Fixed in `1e020a7e`: the help text now says this raises an error at kernel selection on backends that do not declare gluon skip-softmax support, rather than being silently ignored.

File: python/tokenspeed/runtime/utils/server_args.py [hunk 7]
```diff
@@ -1628,6 +1631,23 @@ def add_cli_args(parser: argparse.ArgumentParser):
...
+            "every row in the query tile has exp(block_max_score - "
+            "running_max) below this threshold. 0.0 (default) is exact "
+            "dense attention; the skip rate for a given threshold must be "
+            "calibrated per model and sequence length. Only takes effect on "
+            "prefill requests with no cached prefix; KV-cache-extend paths "
+            "(including FP8/MXFP8 KV cache) never reach this kernel and "
+            "silently ignore the threshold. Raises an error at kernel "
+            "selection if the active attention backend does not declare "
+            "gluon skip-softmax support, rather than being silently "
+            "ignored.",
```
[the remark this answers is not in the document]
Comment (hiroaki-amd) [reply python/tokenspeed/runtime/utils/server_args.py]: Good catch. Fixed in `f2a3cc37`: documented that triton's `Priority.PORTABLE` prefill kernel never clears the planner's postwrite cutoff, so it silently falls through to the extend path like FP8/MXFP8, rather than reaching kernel selection.

File: python/tokenspeed/runtime/utils/server_args.py [hunk 8]
```diff
@@ -1628,6 +1631,23 @@ def add_cli_args(parser: argparse.ArgumentParser):
...
+            type=float,
+            default=ServerArgs.skip_softmax_threshold,
+            help="BLASST skip-softmax sparsity threshold for the gluon MHA "
+            "prefill kernel (gfx950 only). A K/V block is skipped only when "
+            "every row in the query tile has exp(block_max_score - "
+            "running_max) below this threshold. 0.0 (default) is exact "
+            "dense attention; the skip rate for a given threshold must be "
+            "calibrated per model and sequence length. Only takes effect on "
+            "prefill requests with no cached prefix; KV-cache-extend paths "
+            "(including FP8/MXFP8 KV cache) never reach this kernel and "
```
[the remark this answers is not in the document]
Comment (hiroaki-amd) [reply python/tokenspeed/runtime/utils/server_args.py]: Good catch. Fixed in `f2a3cc37`: documented that `max_extend_prefix_len` is a batch-wide max, so one cache-hit request in the batch routes the whole batch through the extend path, not just per-request.

**Review by borontion — DISMISSED** [review]
